src/mon/OSDMonitor.cc

```diff
@@ -74,18 +74,18 @@
 #define OSD_PG_CREATING_PREFIX "osd_pg_creating"
 
 struct C_UpdateCreatingPGs : public Context {
-  OSDMonitor *mon;
+  OSDMonitor *osdmon;
   utime_t start;
   epoch_t epoch;
-  C_UpdateCreatingPGs(OSDMonitor *mon, epoch_t e) :
-    mon(mon), start(ceph_clock_now()), epoch(e) {}
+  C_UpdateCreatingPGs(OSDMonitor *osdmon, epoch_t e) :
+    osdmon(osdmon), start(ceph_clock_now()), epoch(e) {}
   void finish(int r) override {
     if (r >= 0) {
       utime_t end = ceph_clock_now();
       dout(10) << "osdmap epoch " << epoch << " mapping took "
 	       << (end - start) << " seconds" << dendl;
-      mon->update_creating_pgs();
-      mon->check_pg_creates_subs();
+      osdmon->update_creating_pgs();
+      osdmon->check_pg_creates_subs();
```

Comment:
not all of them, the osdmapping thread does not hold `Monitor::lock`. and what complicates this problem is that this completion callback could be called right in `OSDMonitor::start_mapping()` where the `Monitor::lock` is already acquired if the mapping is completed at that moment. so we have two choices:

- change the Monitor::lock to a recursive lock
- add another lock

i think the latter is simpler, and it also avoids the potential problems could be introduced by a single grand lock.